### PR TITLE
Fix | Adding non-string support to SqlConnectionStringBuilder property indexer

### LINKED ISSUE #2015: SqlConnectionStringBuilder property indexer doesn't support non-string values
This code:
```C#
// See https://aka.ms/new-console-template for more information
using Microsoft.Data.SqlClient;

var conn = new SqlConnectionStringBuilder();
try
{
    Console.WriteLine("Set Encrypt to SqlConnectionEncryptionOption type");
    conn["Encrypt"] = SqlConnectionEncryptOption.Strict;
}
catch (Exception e)
{
    Console.WriteLine(e);
}
try
{
    Console.WriteLine("Set Encrypt to boolean false");
    conn["Encrypt"] = false;
}
catch (Exception e)
{
    Console.WriteLine(e);
}
Console.WriteLine(conn.ConnectionString);
```

Prints the following:
```
Set Encrypt to SqlConnectionEncryptionOption type
System.ArgumentException: Invalid value for key 'Encrypt'.
   at Microsoft.Data.Common.DbConnectionStringBuilderUtil.ConvertToSqlConnectionEncryptOption(String keyword, Object value)
   at Microsoft.Data.SqlClient.SqlConnectionStringBuilder.ConvertToSqlConnectionEncryptOption(String keyword, Object value)
   at Microsoft.Data.SqlClient.SqlConnectionStringBuilder.set_Item(String keyword, Object value)
   at Program.<Main>$(String[] args) in E:\git\sqlenc\Program.cs:line 8
Set Encrypt to boolean false
System.ArgumentException: Invalid value for key 'Encrypt'.
   at Microsoft.Data.Common.DbConnectionStringBuilderUtil.ConvertToSqlConnectionEncryptOption(String keyword, Object value)
   at Microsoft.Data.SqlClient.SqlConnectionStringBuilder.ConvertToSqlConnectionEncryptOption(String keyword, Object value)
   at Microsoft.Data.SqlClient.SqlConnectionStringBuilder.set_Item(String keyword, Object value)
   at Program.<Main>$(String[] args) in E:\git\sqlenc\Program.cs:line 17

```
Given that there's client code floating around that uses the indexer to save copies of the properties in a dictionary then writes them back using the indexer in order to capture parameters without knowing their names and types ahead of time, at minimum the indexer should allow one to set `Encrypt` to an instance of `SqlConnectionEncryptionOption`. It'd be nice if it supported a bool setter too.

Otherwise in SSMS 20 we'll have to modify the designer code to special case this option to convert the stored value to a string.

### PR DESCRIPTION
**Pull request by JRahnama**
Fixes #2015

### CODE REVIEW

**Comment by ErikEJ** [comment]
Possible to add a test?

**Comment by JRahnama** [comment]
> Possible to add a test?

Sure, I will add some tests here.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlConnectionEncryptOption.cs [hunk 1]
```diff
@@ -43,6 +43,18 @@ public static SqlConnectionEncryptOption Parse(string value)
...
+        internal static SqlConnectionEncryptOption Parse(bool value)
+        {
+            if (TryParse(value, out SqlConnectionEncryptOption result))
+            {
+                return result;
+            }
+            else
+            {
+                throw ADP.InvalidConnectionOptionValue(SqlConnectionString.KEY.Encrypt);
+            }
```
[the remark this answers is not in the document]
Comment (JRahnama) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlConnectionEncryptOption.cs:55]: the sample repro also fails for bool values.
```C#
using Microsoft.Data.SqlClient;

var conn = new SqlConnectionStringBuilder();
try
{
    Console.WriteLine("Set Encrypt to SqlConnectionEncryptionOption type");
    conn["Encrypt"] = SqlConnectionEncryptOption.Strict;
}
catch (Exception e)
{
    Console.WriteLine(e);
}
try
{
    Console.WriteLine("Set Encrypt to boolean false");
    conn["Encrypt"] = false;
}
catch (Exception e)
{
    Console.WriteLine(e);
}
Console.WriteLine(conn.ConnectionString);
```

File: doc/snippets/Microsoft.Data.SqlClient/SqlConnectionEncryptOption.xml [hunk 2]
```diff
@@ -26,7 +26,7 @@ Implicit conversions have been added to maintain backwards compatibility with bo
     </Parse>
     <TryParse>
       <summary>
-        Converts the specified string representation of a logical value to its <see cref="T:Microsoft.Data.SqlClient.SqlConnectionEncryptOption"/> equivalent and returns a value that indicates whether the conversion succeeded.
+        Converts the specified string, boolean representation of a logical value to its <see cref="T:Microsoft.Data.SqlClient.SqlConnectionEncryptOption"/> equivalent and returns a value that indicates whether the conversion succeeded.
```
[the remark this answers is not in the document]
Comment (JRahnama) [reply doc/snippets/Microsoft.Data.SqlClient/SqlConnectionEncryptOption.xml]: done.

**Review by David-Engel — APPROVED** [review]

File: src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlConnectionStringBuilderTest.cs [hunk 3]
```diff
@@ -402,6 +402,25 @@ public void ConnectionBuilderEncryptBackwardsCompatibility()
...
+                CheckEncryptType(builder, SqlConnectionEncryptOption.Strict);
+                builder["Encrypt"] = SqlConnectionEncryptOption.Optional;
+                CheckEncryptType(builder, SqlConnectionEncryptOption.Optional);
+                builder["Encrypt"] = SqlConnectionEncryptOption.Mandatory;
+                CheckEncryptType(builder, SqlConnectionEncryptOption.Mandatory);
+            }
+            catch (SqlException ex)
+            {
+                Assert.Fail(ex.Message);
+            }
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlConnectionStringBuilderTest.cs:421]: Unexpected exceptions don't need to be handled. Let the exception be thrown with the most information.
```suggestion
            builder["Encrypt"] = SqlConnectionEncryptOption.Strict;
            CheckEncryptType(builder, SqlConnectionEncryptOption.Strict);
            builder["Encrypt"] = SqlConnectionEncryptOption.Optional;
            CheckEncryptType(builder, SqlConnectionEncryptOption.Optional);
            builder["Encrypt"] = SqlConnectionEncryptOption.Mandatory;
            CheckEncryptType(builder, SqlConnectionEncryptOption.Mandatory);
```

File: src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlConnectionStringBuilderTest.cs [hunk 4]
```diff
@@ -413,6 +432,30 @@ public void ConnectionBuilderEncryptBackwardsCompatibility()
...
+                }
+                else
+                {
+                    CheckEncryptType(builder, SqlConnectionEncryptOption.Optional);
+                }
+            }
+            catch (Exception ex)
+            {
+                Assert.Fail(ex.Message);
+            }
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlConnectionStringBuilderTest.cs:456]:
```suggestion
            builder["Encrypt"] = value;
            CheckEncryptType(builder, value ? SqlConnectionEncryptOption.Mandatory : SqlConnectionEncryptOption.Optional);
```

File: src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlConnectionStringBuilderTest.cs [hunk 5]
```diff
@@ -467,5 +510,11 @@ internal void ExecuteConnectionStringTests(string connectionString)
                 Assert.NotNull(connection);
             }
         }
+
+        internal void CheckEncryptType(SqlConnectionStringBuilder builder, SqlConnectionEncryptOption expectedValue)
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlConnectionStringBuilderTest.cs:514]:
```suggestion
        internal static void CheckEncryptType(SqlConnectionStringBuilder builder, SqlConnectionEncryptOption expectedValue)
```

**Review by DavoudEshtehari — APPROVED** [review]

File: src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlConnectionStringBuilderTest.cs [hunk 6]
```diff
@@ -402,6 +402,18 @@ public void ConnectionBuilderEncryptBackwardsCompatibility()
             Assert.True(builder.Encrypt);
         }
 
+        [Fact]
+        public void EncryptParserValidValuesPropertyIndexerForEncryptionOption()
+        {
+            SqlConnectionStringBuilder builder = new();
+            builder["Encrypt"] = SqlConnectionEncryptOption.Strict;
```
[the remark this answers is not in the document]
Comment (JRahnama) [reply src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlConnectionStringBuilderTest.cs:409]: Thanks David. I will update the tests accordingly.

[the remark this answers is not in the document]
Comment (JRahnama) [reply src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlConnectionStringBuilderTest.cs:409]: also this fix is a back port candidate for other related versions, the ones with the change for Encrypt property. dates are TBD.

**Pull request merged**